### DISCUSSION TITLE
Issue with Axios and BigInt Serialization

### DISCUSSION THREAD

**Discussion by kartik-gupta-ij** [axios/axios#6278] [Q&A]
Hello Axios Community,

I hope this message finds you well. I am currently facing an issue related to the serialization of BigInt in Axios, and I'm seeking guidance on how to resolve it.

I've identified that the error is occurring in the following code snippet from Axios:

```javascript
function stringifySafely(rawValue, parser, encoder) {
  // ... (function details)
}

// Source: https://github.com/axios/axios/blob/d844227411263fab39d447442879112f8b0c8de5/lib/defaults/index.js#L21C10-L21C25
```

The error message I'm encountering is:

```
TypeError: Do not know how to serialize a BigInt at stringify
```

As part of my debugging process, I've created custom parser and encoder functions specifically designed for handling BigInts when parsing JSON data. However, I'm struggling to understand how to pass these custom functions to Axios so that they are utilized by the `stringifySafely` function.

Here is a simplified version of my custom functions:

```javascript
let bigintReviver
let bigintReplacer
if ('rawJSON' in JSON) {
  bigintReviver = function (_key, val, context) {
    if (Number.isInteger(val) && !Number.isSafeInteger(val)) {
      return BigInt(context?.source) 
    }
    return val
  }
  bigintReplacer = function (_key, val) {
    if (typeof val === 'bigint') {
      return JSON.rawJSON?.(String(val))
    }
    return val
  }
}

export const jsonBigIntParse = function (jsonString, reviver, parserConfig) {
  if (bigintReviver) {
    return JSON.parse(jsonString, bigintReviver, parserConfig)
  }
  return JSON.parse(jsonString, reviver)
}

export const jsonBigIntParseStringify = function (value, replacer, space) {
  if (bigintReplacer) {
    return JSON.stringify(value, bigintReplacer, space)
  }
  return JSON.stringify(value, replacer, space)
}

```

My question is: How can I properly configure Axios to utilize these custom parser and encoder functions within the `stringifySafely` function? I believe this is the key to resolving the BigInt serialization issue I am facing.

Any guidance or suggestions on the best approach to integrate custom parser and encoder functions with Axios would be highly appreciated.

Thank you in advance for your assistance.

**Comment by jasonsaayman** [comment]
Axios does not expose a way to pass custom `parser` / `encoder` functions directly into `stringifySafely()`. That helper is internal, and the default request transform calls it without custom arguments.

For request bodies, the practical fix is to serialize the payload yourself or provide a custom `transformRequest`.

Example using BigInt as strings, which is the safest JSON-compatible format:

```js
const bigintReplacer = (_key, value) => {
  return typeof value === 'bigint' ? value.toString() : value;
};

await axios.post(
  '/api/example',
  JSON.stringify(data, bigintReplacer),
  {
    headers: {
      'Content-Type': 'application/json'
    }
  }
);
```

Or with transformRequest:

```js
const api = axios.create({
  transformRequest: [function (data, headers) {
    if (data && typeof data === 'object') {
      headers.setContentType('application/json', false);

      return JSON.stringify(data, (_key, value) => {
        return typeof value === 'bigint' ? value.toString() : value;
      });
    }

    return data;
  }]
});
```

For response parsing, recent Axios versions support parseReviver, which is passed to `JSON.parse()`:

```js
const response = await axios.get('/api/example', {
  parseReviver(key, value, context) {
    if (
      typeof value === 'number' &&
      !Number.isSafeInteger(value) &&
      /^-?\d+$/.test(context?.source || '')
    ) {
      return BigInt(context.source);
    }

    return value;
  }
});
```

If your Axios version does not support parseReviver, use transformResponse and parse the response yourself:

```js
const response = await axios.get('/api/example', {
  transformResponse: [function (data) {
    return JSON.parse(data, (key, value) => {
      // Convert based on your API contract, for example specific fields:
      if (key === 'id' && typeof value === 'string' && /^\d+$/.test(value)) {
        return BigInt(value);
      }

      return value;
    });
  }]
});
```

One important limitation: standard JSON has no native BigInt type. If the server sends large integers as JSON numbers, older runtimes may already lose precision before your reviver can recover them. The most reliable API format is to send large integers as strings and convert them back to BigInt based on known field names or schema.
